File: nixpkgut/nardl.go
package nixpkgut

import (
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

type DownloadOptions struct {
	StorePath  string
	Output    string
	CacheURL  string
}

var defaultCacheURL = "https://cache.nixos.org"

func init() {
	if url := os.Getenv("NIX_CACHE_URL"); url != "" {
		defaultCacheURL = url
	}
}

func getNarURL(cacheURL, storePath string) (narinfoURL, narURL string, err error) {
	hash := strings.TrimPrefix(storePath, "/nix/store/")
	hash = strings.SplitN(hash, "-", 1)[0]
	hashpart := strings.Split(hash, "-")[0]

	hash = strings.TrimPrefix(storePath, "/nix/store/")
	parts := strings.SplitN(hash, "-", 2)
	if len(parts) < 2 {
		return "", "", fmt.Errorf("invalid store path: %s", storePath)
	}
	hashpart = parts[0]

	narinfoURL = fmt.Sprintf("%s/%s.narinfo", cacheURL, hashpart)
	narURL = fmt.Sprintf("%s/nar/%s.nar.xz", cacheURL, hashpart)

	return narinfoURL, narURL, nil
}

func downloadNar(opts *DownloadOptions) error {
	cacheURL := opts.CacheURL
	if cacheURL == "" {
		cacheURL = defaultCacheURL
	}

	storePath := opts.StorePath
	output := opts.Output

	if storePath == "" {
		flag.Usage()
		os.Exit(1)
	}

	_, narURL, err := getNarURL(cacheURL, storePath)
	if err != nil {
		return err
	}

	req, err := http.NewRequest("GET", narURL, nil)
	if err != nil {
		return err
	}

	client := &http.Client{Timeout: 60 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	if output == "" || output == "-" {
		os.Stdout.Write(body)
	} else {
		err = os.WriteFile(output, body, 0644)
		if err != nil {
			return err
		}
		fmt.Printf("Saved to %s\n", output)
	}

	return nil
}

func RunDownload(args []string) error {
	opts := &DownloadOptions{}

	flag.StringVar(&opts.StorePath, "s", "", "Store path (e.g., /nix/store/xxx-hello-2.12.3)")
	flag.StringVar(&opts.Output, "o", "", "Output file (default: stdout)")
	flag.StringVar(&opts.CacheURL, "c", "", "Cache URL (default: https://cache.nixos.org)")

	flag.Parse()

	parsedArgs := flag.Args()
	if len(parsedArgs) > 0 {
		opts.StorePath = parsedArgs[0]
	}

	if opts.StorePath == "" {
		flag.Usage()
		return fmt.Errorf("store path required")
	}

	return downloadNar(opts)
}

func main() {
	if err := Run(os.Args[1:]); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}

File: nixpkgut/nardl.c
#include "nardl.h"
#include "nixse.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <curl/curl.h>

#define DEFAULT_CACHE_URL "https://cache.nixos.org"

typedef struct {
    char *data;
    size_t size;
    FILE *fp;
} DownloadContext;

static size_t write_file_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    size_t realsize = size * nmemb;
    DownloadContext *ctx = (DownloadContext *)userp;
    
    if (ctx->fp) {
        return fwrite(contents, size, nmemb, ctx->fp);
    }
    
    char *ptr = realloc(ctx->data, ctx->size + realsize + 1);
    if (!ptr) return 0;
    
    ctx->data = ptr;
    memcpy(ctx->data + ctx->size, contents, realsize);
    ctx->size += realsize;
    ctx->data[ctx->size] = '\0';
    
    return realsize;
}

static void get_hash_from_store_path(const char *store_path, char *hash_out, size_t hash_size) {
    const char *prefix = "/nix/store/";
    if (strncmp(store_path, prefix, strlen(prefix)) == 0) {
        const char *hash_start = store_path + strlen(prefix);
        const char *dash = strchr(hash_start, '-');
        if (dash && (size_t)(dash - hash_start) < hash_size) {
            strncpy(hash_out, hash_start, dash - hash_start);
            hash_out[dash - hash_start] = '\0';
            return;
        }
    }
    hash_out[0] = '\0';
}

int download_nar(const DownloadOptions *opts) {
    const char *cache_url = opts->cache_url ? opts->cache_url : DEFAULT_CACHE_URL;
    const char *store_path = opts->store_path;
    const char *output = opts->output;
    
    if (!store_path) {
        fprintf(stderr, "Error: store path required\n");
        return 1;
    }
    
    char hash[64];
    get_hash_from_store_path(store_path, hash, sizeof(hash));
    
    if (strlen(hash) == 0) {
        fprintf(stderr, "Error: invalid store path\n");
        return 1;
    }
    
    char nar_url[512];
    snprintf(nar_url, sizeof(nar_url), "%s/nar/%s.nar.xz", cache_url, hash);
    
    CURL *curl = curl_easy_init();
    if (!curl) {
        fprintf(stderr, "Error: curl init failed\n");
        return 1;
    }
    
    DownloadContext ctx = {0};
    
    if (output && strcmp(output, "-") != 0) {
        ctx.fp = fopen(output, "wb");
        if (!ctx.fp) {
            fprintf(stderr, "Error: cannot open output file: %s\n", output);
            curl_easy_cleanup(curl);
            return 1;
        }
    }
    
    curl_easy_setopt(curl, CURLOPT_URL, nar_url);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_file_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &ctx);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 60L);
    
    CURLcode res = curl_easy_perform(curl);
    
    long http_code = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
    
    if (ctx.fp) fclose(ctx.fp);
    free(ctx.data);
    curl_easy_cleanup(curl);
    
    if (res != CURLE_OK) {
        fprintf(stderr, "Error: download failed: %s\n", curl_easy_strerror(res));
        return 1;
    }
    
    if (http_code != 200) {
        fprintf(stderr, "Error: HTTP %ld\n", http_code);
        return 1;
    }
    
    if (output && strcmp(output, "-") != 0) {
        printf("Saved to %s\n", output);
    }
    
    return 0;
}